cilium/CFP-28986-mutual-auth-updates.md

```diff
@@ -0,0 +1,256 @@
+
+## Proposal
+
+### Overview
+
+This CFP proposes two changes: enforcing every packet contains an identity using either encryption or encapsulation and adding a per-connection mutual auth mode. These changes are relatively orthogonal and could be implemented as two separate changes.
+
+#### Requiring encryption or encapsulation resolves cache manipulation
+
+This assertion depends on two things that may not be clear:
+
+- Wireguard encryption now (as of Cilium 1.15) also encapsulates packets into VXLAN as they transit the tunnel, adding a Cilium identity into the header (as in unencrypted VXLAN)
+- When the datapath can determine the remote security identity from packets directly, the ipcache is _not_ used.
+
+As part of actions taken to mitigate this attack, we added VXLAN encapsulation to Wireguard during development of Cilium 1.15.
+
+But Cilium's eBPF code has _always_ preferenced identity information it decodes from encapsulation, and in the case that it's present, the ipcache is not used for policy decisions.
+
+This is why the original attack required the use of direct routing - when encapsulation is in use, the attack never worked.
+
+The more recent change to ensure that Wireguard encryption also includes encapsulation just means that when you're ensuring your traffic's wire security, you're also making the identity not spoofable by cache manipulation.
+
+#### Diagrams of the new workflow
+
+After adding this change, let's look at diagrams of before and after the exploit is attempted:
+
+![Node state before cache manipulation exploit, with identity in packet](images/28986-before-exploit-with-auth-and-encryption.png)
+
+Before, things work as anticipated.
+
+However, it's after that there is a difference:
+
+![Node state after cache manipulation exploit, with identity in packet](images/28986-after-exploit-with-auth-and-encryption.png)
+
+Even though the ipcache entries don't match on both nodes, the identity being included in the packets means that the datapath relies on that instead of the ipcache, so any cache manipulation is irrelevant.
+
+### Connection-based mutual authentication method
+
+The connection-based mutual authentication method will:
+* Be selectable as an entirely new authentication method (called `connection-spiffe` or similar)
+* Use the conntrack table to keep track of connections
+* Add two bits of marking into conntrack table entries, "Auth required", and "Auth completed"
+* Use the existing TLS handshake code, but call it on a different event (new connection instead of new identity-identity-node tuple).
+
+The existing conntrack table performs standard connection tracking functions, tracking connections based on the 5-tuple of source IP, source port, destination IP, destination port, and protocol. TCP is handled using its connection mechanisms, and UDP is handled using the same time-based pseudo connection that is standard across most networking gear.
```

Comment:
I don't know enough about UDP pseudo connection, but if we do per-connection auth for UDP, what are the chances that we drop so many packets during mutual auth that the UDP traffic is ineffective?

Reply:
Reasonable, depending on how the application using UDP treats its UDP packets. Many UDP applications essentially build their own connection handling on top of UDP (HTTP3 is a good example).

A UDP pseudo-connection is created by most connection-tracking tools by associating UDP packets matching a 5-tuple (and the equivalent return traffic) into a "connection", even though it's only a guess that they actually are a connection. These pseudo-connections usually have a timer that has them be removed once no packets have matched for a certain period of time.

What this means in this case is that UDP traffic will also be captured by the conntrack table, even though there's not actually any connection involved.